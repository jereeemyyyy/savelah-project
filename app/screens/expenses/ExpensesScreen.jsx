import { Text, View } from 'react-native';

export default function ExpensesScreen() {
    return (
        <View>
            <Text className="font-bold text-xl">Expenses Screen!</Text>
        </View>
    );
}