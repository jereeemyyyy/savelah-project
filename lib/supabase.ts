import { AppState } from 'react-native'
import 'react-native-url-polyfill/auto'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = 'https://cjcldaiiqmedjmojjzfz.supabase.co'
const supabaseAnonKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNqY2xkYWlpcW1lZGptb2pqemZ6Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MTY3OTE2NDksImV4cCI6MjAzMjM2NzY0OX0.6z2M2mG_qmSPPwujMWYpN4KkLgiU0mWjGdsCdVZvlNw'

export const supabase = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    storage: AsyncStorage,
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: false,
  },
})

// Tells Supabase Auth to continuously refresh the session automatically
// if the app is in the foreground. When this is added, you will continue
// to receive `onAuthStateChange` events with the `TOKEN_REFRESHED` or
// `SIGNED_OUT` event if the user's session is terminated. This should
// only be registered once.
AppState.addEventListener('change', (state) => {
  if (state === 'active') {
    supabase.auth.startAutoRefresh()
  } else {
    supabase.auth.stopAutoRefresh()
  }
})