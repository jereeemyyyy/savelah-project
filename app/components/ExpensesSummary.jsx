import React, { useState, useEffect } from 'react';
import { View, Text, Dimensions, SafeAreaView, TouchableOpacity } from 'react-native';
import { PieChart } from 'react-native-chart-kit';
import { supabase } from '../../lib/supabase';
import getColour from '../utils/getColour';

const screenWidth = Dimensions.get('window').width;

export default function ExpensesSummary({ userId }) {
  const [data, setData] = useState([]);
  const [currentWeekTotal, setCurrentWeekTotal] = useState(0);
  const [lastWeekTotal, setLastWeekTotal] = useState(0);

  const fetchExpensesByCategory = async () => {
    try {
      const { data: fetchedData, error } = await supabase.rpc('get_total_expenses_by_category', {
        p_user_id: userId
      });

      if (error) {
        console.log(error);
        throw new Error('Error fetching expenses by category');
      }

      // Map the fetched data to the format expected by PieChart and validate it
      const formattedData = fetchedData.map(item => {
        const amount = parseFloat(item.total_expense);
        const color = getColour(item.category);

        return {
          name: item.category,
          population: item.total_expense,
          color,
          legendFontColor: '#000000',
          legendFontSize: 12
        };
      }).filter(item => item !== null); // Remove invalid entries

      setData(formattedData);
    } catch (error) {
      console.error('Error fetching expenses by category:', error.message);
    }
  };

  const fetchWeeklyTotals = async () => {
    try {
      const { data: currentWeekData, error: currentWeekError } = await supabase.rpc('get_total_expenses_current_week', {
        p_user_id: userId
      });

      if (currentWeekError) {
        console.error('Error fetching current week expenses:', currentWeekError);
        return;
      }

      setCurrentWeekTotal(currentWeekData);

      const { data: lastWeekData, error: lastWeekError } = await supabase.rpc('get_total_expenses_last_week', {
        p_user_id: userId
      });

      if (lastWeekError) {
        console.error('Error fetching last week expenses:', lastWeekError);
        return;
      }

      setLastWeekTotal(lastWeekData);
    } catch (error) {
      console.error('Error fetching weekly totals:', error.message);
    }
  };

  useEffect(() => {
    if (userId) {
      fetchExpensesByCategory();
      fetchWeeklyTotals();
    }

    const subscription = supabase
      .channel('room1')
      .on('postgres_changes', { event: '*', schema: 'public', table: 'categories', filter: `user_id=eq.${userId}` }, payload => {
        fetchExpensesByCategory();
        fetchWeeklyTotals();
      })
      .subscribe();

    return () => {
      supabase.removeChannel(subscription);
    };
  }, [userId]);

  const handleRefresh = () => {
    fetchExpensesByCategory();
    fetchWeeklyTotals();
  };

  return (
    <SafeAreaView className="flex-1 justify-start items-center">
      <TouchableOpacity className="bg-white rounded-lg p-5 shadow-lg w-11/12" onPress={handleRefresh}>
        {data.length > 0 ? (
          <PieChart
            data={data.map(item => ({
              name: item.name,
              population: item.population,
              color: item.color,
              legendFontColor: item.legendFontColor,
              legendFontSize: item.legendFontSize
            }))}
            width={screenWidth - 80}
            height={150}
            chartConfig={{
              backgroundColor: '#ffffff',
              backgroundGradientFrom: '#ffffff',
              backgroundGradientTo: '#ffffff',
              color: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
            }}
            accessor="population"
            backgroundColor="transparent"
            paddingLeft="10"
            paddingRight="0"
            absolute
          />
        ) : (
          <Text className="text-gray-600 mb-4">No expenses data available.</Text>
        )}

        <View className="mt-1">
          <Text className="text-base text-black font-bold">Last Week: ${lastWeekTotal}</Text>
          <Text className="text-base text-black font-bold">This Week: ${currentWeekTotal}</Text>
        </View>
      </TouchableOpacity>
    </SafeAreaView>
  );
}
