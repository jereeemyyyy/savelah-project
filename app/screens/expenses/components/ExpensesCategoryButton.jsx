// App.js
import React, { useState, useEffect } from 'react';
import { View, Text, TouchableOpacity, FlatList, TextInput, Button, Modal } from 'react-native';
import { supabase } from '../../../../lib/supabase';
import ExpensesModal from './ExpensesModal';
import Icon from 'react-native-vector-icons/FontAwesome';


// Initial categories with predefined colors
const initialCategories = {
  Food: '#FF0000',            // Red
  Transportation: '#0000FF',  // Blue
  Housing: '#00FF00',         // Green
  Others: '#FFFF00',          // Yellow
};

// Function to generate a random color
const getRandomColor = () => {
  const letters = '0123456789ABCDEF';
  let color = '#';
  for (let i = 0; i < 6; i++) {
    color += letters[Math.floor(Math.random() * 16)];
  }
  return color;
};

export default function ExpensesCategoryButton({userId}) {

  const [categories, setCategories] = useState([]);
  const [modalVisible, setModalVisible] = useState(false);
  const [selectedItem, setSelectedItem] = useState(null);
  

  useEffect(() => {
    fetchCategoriesAndExpenses();

    const subscription = supabase
      .channel('room1')
      .on('postgres_changes', 
        { event: '*', 
          schema: 'public', 
          table: 'categories', 
          filter: `user_id=eq.${userId}` 
        }, payload => {
          fetchCategoriesAndExpenses(); 
        })
        .subscribe();

    return () => {
      supabase.removeChannel(subscription);
    };
  }, [userId]);

  const fetchCategoriesAndExpenses = async () => {
    try {
      const { data, error } = await supabase.rpc('get_total_expenses_by_category', { p_user_id: userId });
      
      if (error) {
        console.error('Error fetching data:', error);
        throw error;
      }
      
      setCategories(data);
    } catch (error) {
      console.error('Error in fetchCategoriesAndExpenses:', error.message);
    }
  };


  const handlePressItem = (item) => {
    setSelectedItem(item);
    setModalVisible(true);
  };

  const renderItem = ({ item }) => (
    <TouchableOpacity className="w-1/2 p-2" onPress={() => handlePressItem(item)}>
      <View className="bg-gray-200 p-4 rounded-lg">
        <Text className="text-xl font-bold">${item.total_expense}</Text>
        <Text className="text-gray-600">{item.category}</Text>
        <View style={{ height: 4, marginTop: 8, backgroundColor: getRandomColor() }} />
      </View>
    </TouchableOpacity>
  );

  return (
    <View className="flex-1 p-4 bg-gray-800">
      <View className="flex-row items-center justify-between">
        <Text className="font-bold text-3xl p-4 text-white -ml-2 -mt-2">Categories</Text> 
        <TouchableOpacity
            onPress={fetchCategoriesAndExpenses}
            className="bg-purple-500 p-3 rounded ml-4 mr-2"
          >
            <Icon name="refresh" size={22} color="white" />
          </TouchableOpacity>
      </View>

      {categories.length <= 0 ? (
        <View className="flex-1 justify-center items-center">
          <Text className="text-white text-lg">No Categories available.</Text>
          <Text className="text-white text-lg mb-8">Create one in the Budgets screen!</Text>
        </View>
      ) : (
        <FlatList
          data={categories}
          renderItem={({ item }) => renderItem({ item })}
          keyExtractor={(item) => item.category.toString()}
          numColumns={2}
          />
      )}
      
      
      {selectedItem && (
        <ExpensesModal 
          selectedItem={selectedItem} 
          visible={modalVisible} 
          onClose={() => setModalVisible(!modalVisible)}
        />
      )}
    </View>
  );
}
