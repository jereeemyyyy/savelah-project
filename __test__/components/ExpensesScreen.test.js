import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { Text, View } from 'react-native';
import ExpensesScreen from '../../app/screens/expenses/ExpensesScreen';
import { supabase } from '../../lib/supabase';
import ExpensesSummary from '../../app/components/ExpensesSummary';
import ExpensesCategoryButton from '../../app/screens/expenses/components/ExpensesCategoryButton';

jest.mock('../../lib/supabase', () => ({
    supabase: {
        auth: {
            getUser: jest.fn(),
        },
    },
}));

jest.mock('../../app/components/ExpensesSummary', () => 'ExpensesSummary');
jest.mock('../../app/screens/expenses/components/ExpensesCategoryButton', () => 'ExpensesCategoryButton');

describe('ExpensesScreen', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });

        let tree;
        await act(async () => {
            tree = renderer.create(<ExpensesScreen />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('contains the expected text elements', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });

        let component;
        await act(async () => {
            component = renderer.create(<ExpensesScreen />);
        });
        const instance = component.root;

        expect(instance.findByType(Text).props.children).toContain('Expenses This Week');
    });

    it('contains the ExpensesSummary and ExpensesCategoryButton components', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });

        let component;
        await act(async () => {
            component = renderer.create(<ExpensesScreen />);
        });
        const instance = component.root;

        expect(instance.findByType('ExpensesSummary')).toBeTruthy();
        expect(instance.findByType('ExpensesCategoryButton')).toBeTruthy();
    });
});
