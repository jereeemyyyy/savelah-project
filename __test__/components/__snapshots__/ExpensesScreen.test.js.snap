// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpensesScreen renders correctly 1`] = `
<View
  style={
    [
      {
        "justifyContent": "flex-start",
      },
      {
        "backgroundColor": "#1f2937",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "flexDirection": "column",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "fontWeight": "700",
        },
        {
          "fontSize": 30,
          "lineHeight": 36,
        },
        {
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
        {
          "color": "#fff",
        },
      ]
    }
  >
    Expenses This Week
  </Text>
  <View
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "marginBottom": -40,
        },
      ]
    }
  >
    <ExpensesSummary
      userId="123"
    />
  </View>
  <View
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "marginTop": -80,
        },
      ]
    }
  >
    <ExpensesCategoryButton
      userId="123"
    />
  </View>
</View>
`;
